Spigot-Server-Patches/0446-Implement-Mob-Goal-API.patch

```diff
@@ -0,0 +1,446 @@
++        <dependency>
++            <groupId>io.github.classgraph</groupId>
++            <artifactId>classgraph</artifactId>
++            <version>4.8.47</version>
++            <scope>test</scope>
++        </dependency>
+     </dependencies>
+ 
+     <repositories>
+diff --git a/src/main/java/com/destroystokyo/paper/ai/PaperCustomGoal.java b/src/main/java/com/destroystokyo/paper/ai/PaperCustomGoal.java
+new file mode 100644
+index 00000000..edec59de
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/ai/PaperCustomGoal.java
+@@ -0,0 +1,44 @@
++package com.destroystokyo.paper.ai;
++
++import net.minecraft.server.PathfinderGoal;
++
++/**
++ * Wraps api in vanilla
++ */
++public class PaperCustomGoal extends PathfinderGoal {
++
++    private Goal handle;
++
++    public PaperCustomGoal(Goal handle) {
++        this.handle = handle;
++    }
++
++    @Override
++    public boolean shouldActivate() {
++        return handle.shouldActivate();
++    }
++
++    @Override
++    public boolean shouldStayActive() {
++        return handle.shouldStayActive();
++    }
++
++    @Override
++    public void start() {
++        handle.start();
++    }
++
++    @Override
++    public void onTaskReset() {
++        handle.stop();
++    }
++
++    @Override
++    public void tick() {
++        handle.tick();
++    }
++
++    public Goal getHandle() {
++        return handle;
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/ai/PaperGoalSelector.java b/src/main/java/com/destroystokyo/paper/ai/PaperGoalSelector.java
+new file mode 100644
+index 00000000..378c82c7
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/ai/PaperGoalSelector.java
+@@ -0,0 +1,85 @@
++package com.destroystokyo.paper.ai;
++
++import net.minecraft.server.PathfinderGoalSelector;
++
++import java.util.Collection;
++import java.util.HashSet;
++import java.util.Set;
++
++import org.bukkit.NamespacedKey;
++
++public class PaperGoalSelector implements GoalSelector {
++
++    private PathfinderGoalSelector handle;
++
++    public PaperGoalSelector(PathfinderGoalSelector handle) {
++        this.handle = handle;
++    }
++
++    @Override
++    public void addGoal(int priority, Goal goal) {
++        handle.addGoal(priority, new PaperCustomGoal(goal));
++    }
++
++    @Override
++    public void removeGoal(Goal goal) {
++        removeGoal(goal.getKey());
++    }
++
++    @Override
++    public void removeAllGoals() {
++        for (Goal goal : getAllGoals()) {
++            removeGoal(goal);
++        }
++    }
++
++    @Override
++    public void removeGoal(NamespacedKey key) {
++        for (Goal g : getAllGoals()) {
++            if (g.getKey().equals(key)) {
++                handle.removeGoal(((PaperVanillaGoal) g).getHandle());
++                break;
++            }
++        }
++    }
++
++    @Override
++    public boolean hasGoal(NamespacedKey key) {
++        for (Goal g : getAllGoals()) {
++            if (g.getKey().equals(key)) {
```

Comment:
invoking all the instance creation below for this operation seems wasteful, should just be able to port the below logic w/o any wrapping?

Reply:
there are no keys in vanilla, I introduced that concept. I auto-generate keys for vanilla goals in the wrapper.

Reply:
even more reason to try to avoid creating every call and store a single instance, to avoid that heavy logic.